Call MPSAllocator callbacks when allocation fails.

dfalbel opened the pull request:
Fixes #87374

@kulinseth and @albanD This makes the MPSAllocator call the MPSAllocatorCallbacks when getting a free buffer and a first try on allocating fails. User can register callbacks that might free a few buffers and an allocation will be retried.

The reason why we need the `recursive_mutex` is that since callbacks are supposed to free memory, they will eventually call free_buffer() that will lock the same `mutex` that's used for allocation. This approach is similar what's used with the `FreeMemoryCallback` in the `CUDACachingAllocator`.

This PR tries to be as minimal as possible, but there could be some additional improvements cleanups, like:

- In current main, there's no way callbacks can be called, so we could probably rename the callback registry to something reflect the same naming in the CudaAllocator:

https://github.com/pytorch/pytorch/blob/996cc1c0d09a7bc6ad33441c08961226005c69bf/c10/cuda/CUDACachingAllocator.h#L14-L24

- Review the EventTypes here:

https://github.com/pytorch/pytorch/blob/996cc1c0d09a7bc6ad33441c08961226005c69bf/aten/src/ATen/mps/MPSAllocator.h#L18-L23

- And IMHO a nice improvement would be if callbacks could be aware of AllocParams, so they can decide to be more agressive or not depending on how much memory is requested. So I'd pass AllocParams in the signature of the executeCallback instance:

https://github.com/pytorch/pytorch/blob/996cc1c0d09a7bc6ad33441c08961226005c69bf/aten/src/ATen/mps/MPSAllocator.h#L25

Let me know if you think we could sneak those changes into this PR or if it's better to propose them in other smaller PR's.

kulinseth commented:
@dfalbel , thanks for starting the PR. The change overall looks good to me. I have added @razarmehr  for any further feedback.

kulinseth commented:
@dfalbel , can you please add a Testcase or a Network which exposes the problem for needing the `recursive_mutex`.

kulinseth commented on aten/src/ATen/mps/MPSAllocator.mm:
```diff
@@ -202,6 +212,9 @@
         alloc_buffer(params) ||
+        // Callbacks might release more memory (eg. by forcing a GC in the host language) thus
+        // we can retry getting a free buffer in the pool, before trying to alloc again.
+        (trigger_callbacks() && get_free_buffer(params)) ||
```
I think we should separate out the change of `recursive_mutex` with the Release event type.

dfalbel replied:
I changed them to be in 2 different commits.

razarmehr commented on aten/src/ATen/mps/MPSAllocator.mm:
```diff
@@ -202,6 +212,9 @@
         alloc_buffer(params) ||
+        // Callbacks might release more memory (eg. by forcing a GC in the host language) thus
+        // we can retry getting a free buffer in the pool, before trying to alloc again.
+        (trigger_callbacks() && get_free_buffer(params)) ||
```
Also, if you're intention was to know "when `get_free_buffer()` fails (due to buffers not available in pool), only then send a callback", then your order of function calling is wrong. You should change it to this instead, otherwise it'll [short-circuit](https://en.wikipedia.org/wiki/Short-circuit_evaluation):
`(get_free_buffer(params) && trigger_callbacks())`

razarmehr commented:
Hi @dfalbel, thank you for your comments for improvements.
1-  Regarding the renaming of registry and event types: This callback is not meant to be for "releasing" the buffers only. We've reserved that callback and event types (for future use) to be more generic, so it could also be triggered when buffer gets allocated, freed back to pool, released in memory, or recycled out of pool.

2-  `AllocParams` is an internal structure. If user needs more information in the callback, we could add them as arguments to  `executeMPSAllocatorCallback()` (or as a single arg by defining a new struct in `IMpsAllocatorCallback`).
The better alternative is to query the `buffer` pointer that is passed along `executeMPSAllocatorCallback(void* buffer)`, and get info about the buffer like this sample code (gets unaligned buffer size) using the new [MPSAllocatorInterface.h](https://github.com/pytorch/pytorch/pull/94327):
```
#include <ATen/mps/MPSAllocatorInterface.h>
...
class SomeAllocatorCallback : public IMpsAllocatorCallback {
...
void executeMPSAllocatorCallback(void* ptr, EventType event) override { 
   if (event == EventType::[some desired event]) {
     size_t buffer_size = getIMPSAllocator()->getUnalignedBufferSize(ptr);
     ...
  }
}
...
```

dfalbel committed:
Make a `std::recursive_mutex`.

This allows callbacks to make use of the allocator. For instance, a callback might be able to 'free' some allocated memory to allow the a `malloc` to succeed.

dfalbel committed:
Trigger memory callbacks when allocation fails

Memory callbacks are executed when getting a buffer from the pool and allocating more memory from the OS fails. They are executed with a new`EventType::ALLOCATION_FAILED`.

dfalbel committed:
MPSAllocator unit test

This unit test allow making sure that Callbacks are correctly called when allocation fails. It also shows that the `recursive_mutex` is necessary.

dfalbel commented on aten/src/ATen/mps/MPSAllocator.mm:
```diff
@@ -202,6 +212,9 @@
         alloc_buffer(params) ||
+        // Callbacks might release more memory (eg. by forcing a GC in the host language) thus
+        // we can retry getting a free buffer in the pool, before trying to alloc again.
+        (trigger_callbacks() && get_free_buffer(params)) ||
```
Agreed! Made the changes.

dfalbel commented on aten/src/ATen/mps/MPSAllocator.mm:
```diff
@@ -179,6 +179,16 @@
   return true;
 }
 
+bool trigger_callbacks () {
```
Ohh I missed it, I just changed the code to use it instead.

dfalbel commented:
Thanks @razarmehr and @kulinseth for the review. I think I have addressed the requested changes.

@kulinseth I added a unit test that exposes the need for the `recursive_mutex` and also tests that the callback is correctly called. However since the callback is only called after failing allocations, the test takes ~5s. Let me know if that's OK, or if I should move it to a different place.

razarmehr reviewed (approved):
LGTM

razarmehr commented:
@pytorchbot merge -f "Lint and MPS tests are green"

albanD reviewed (approved):
SGTM

razarmehr commented:
@pytorchbot merge -f "Lint and MPS tests are green"
